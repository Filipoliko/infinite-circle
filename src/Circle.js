import uuid from './uuid';

export default class Circle {
  constructor({ listen, unlisten, filter, execute } = {}) {
    this.id = uuid();

    this._listen = listen || (() => {});
    this._unlisten = unlisten || (() => {});
    this._filter = filter || this._defaultFilter;
    this._execute = execute || ((...args) => this._defaultExecute(...args));

    this._entries = new Map();

    this._observers = [];

    this._notifyWithContext = this.notify.bind(this);
  }

  subscribe(listener) {
    this._observers.push(listener);
  }

  notify(...args) {
    this._observers.map(listener => {
      listener(...args);
    });
  }

  getEntries() {
    return this._entries;
  }

  register({ read, write, options = { interval: 180 } }) {
    let id = uuid();

    this._entries.set(id, {
      read,
      write,
      options,
      lastExecution: Date.now() - options.interval - 1,
      lastArgs: []
    });

    if (this._entries.size === 1) {
      this._listen(this._notifyWithContext);
    }

    return id;
  }

  unregister(id) {
    this._entries.delete(id);

    if (this._entries.size === 0) {
      this._unlisten(this._notifyWithContext);
    }
  }

  filter(circleEntries, ...args) {
    return this._filter(circleEntries, ...args);
  }

  read(circleEntries, ...args) {
    const now = Date.now();

    return circleEntries.map(entry => {
      entry.lastExecution = now;
      entry.lastArgs = args;
      let readedValue =
        typeof entry.read === 'function' ? entry.read(...args) : null;

      return Object.assign({ readedValue }, entry);
    });
  }

  write(circleEntries, ...args) {
    return circleEntries.map(entry => {
      if (typeof entry.write === 'function') {
        entry.write(entry.readedValue, ...args);
      }
    });
  }

  *execute(...args) {
    if (this._lockCircle || this._entries.size === 0) {
      return [];
    }

    this._lockCircle = true;
    let modifiedEntries = yield* this._execute(...args);
    this._lockCircle = false;

    let isCallAllEntriesWithArgs = !!Array.from(this._entries.values()).filter(
      entry => entry.lastArgs !== args
    ).length;

    if (!isCallAllEntriesWithArgs) {
      this.notify(...args);
    }

    return modifiedEntries;
  }

  _defaultFilter(circleEntries) {
    // if (args.length === 0 ) {
    //     return [];
    // }

    const now = Date.now();

    return Array.from(circleEntries.values()).filter(loopEntry => {
      return loopEntry.lastExecution + loopEntry.options.interval <= now;
    });
  }

  *_defaultExecute(...args) {
    let circleEntries = this.getEntries();

    circleEntries = this.filter(circleEntries, ...args);
    circleEntries = (yield circleEntries) || circleEntries;
    circleEntries = this.read(circleEntries, ...args);
    circleEntries = (yield circleEntries) || circleEntries;
    circleEntries = this.write(circleEntries, ...args);

    return circleEntries;
  }
}
