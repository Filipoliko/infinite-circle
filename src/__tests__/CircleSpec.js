import Circle from '../Circle';
import { toMock } from 'to-mock';

describe('Circle', () => {
  let circle = null;

  const entry = {
    read() {},
    write() {},
    options: {
      interval: 60
    }
  };

  const args = { event: 'event' };
  const readedValue = 111;

  const NativeDate = Date;
  const MockedDate = toMock(Date);

  beforeEach(() => {
    spyOn(MockedDate, 'now').and.returnValue(1);

    circle = new Circle();
    Date = MockedDate; // eslint-disable-line no-global-assign
  });

  afterEach(() => {
    Date = NativeDate; // eslint-disable-line no-global-assign
  });

  it('should call listen method for registering first entry', () => {
    spyOn(circle, '_listen');

    circle.register(entry);

    expect(circle._listen).toHaveBeenCalled();
    expect(circle.getEntries().size).toEqual(1);
  });

  it('should call unlisten method for unregistering last entry', () => {
    spyOn(circle, '_unlisten');

    const id = circle.register(entry);
    circle.unregister(id);

    expect(circle._unlisten).toHaveBeenCalled();
    expect(circle.getEntries().size).toEqual(0);
  });

  it('should call read method which return right structure for write method', () => {
    spyOn(entry, 'read').and.returnValue(readedValue);

    circle.register(entry);
    const entries = Array.from(circle.getEntries().values());

    expect(circle.read(entries, args)).toMatchSnapshot();
    expect(entry.read).toHaveBeenCalledWith(args);
  });

  it('should call read method which return right structure for write method', () => {
    spyOn(entry, 'write');

    circle.register(entry);
    const entries = Array.from(circle.getEntries().values()).map(entry =>
      Object.assign({ readedValue }, entry)
    );

    circle.write(entries, args);

    expect(entry.write).toHaveBeenCalledWith(readedValue, args);
  });
});
