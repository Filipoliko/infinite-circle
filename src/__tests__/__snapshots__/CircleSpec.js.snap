// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle should call read method which return right structure for write method 1`] = `
Array [
  Object {
    "lastArgs": Array [
      Object {
        "event": "event",
      },
    ],
    "lastExecution": 1,
    "options": Object {
      "interval": 60,
    },
    "read": [Function],
    "readedValue": 111,
    "write": [Function],
  },
]
`;
